### PR TITLE
First mongodb protocol support proposal

### PR DESCRIPTION
**Pull request by albanm**
Hello,

Everything is in the commit name: First mongodb protocol support proposal

See https://github.com/elastic/packetbeat/issues/67

This is not really finished. I still need to add support for a few options and improve test coverage. But mostly it works. I have kep a small TODO list in [protos/mongodb/README.md](protos/mongodb/README.md).

I am a go newbie and still unclear about a few things (error management for example), so I would be very glad for someone to review this. I am willing to work on remarks in order to make it clean.

### CODE REVIEW

**Comment by tsg** [comment]
Thanks for the great work. At first glance it looks very good. We'll review in detail over the next few days.

**Comment by tsg** [comment]
@albanm May I ask you to sign the contributor's agreement, please? https://www.elastic.co/contributor-agreement/

**Comment by tsg** [comment]
This is a bit difficult to merge because we did some code reorganization in 1.0.0-beta1, especially splitting some parts in libbeat. 

The way I'd like to proceed with this is:

* I merge it in a 'mongodb' branch first
* I rebase it to master, solve conflicts and make test work
* Optionally, we add a few more tests
* Review again and merge into master

@albanm, does this sound good to you?

**Comment by albanm** [comment]
I just signed the contributor agreement.

Yes, sounds good to me. Tell me when you are done rebasing and merging, I will fetch this new branch so that I can work on the remaining todos and on your comments.

**Comment by tsg** [comment]
Thanks @albanm.

I merged into the mongodb branch and update it to the changes in Packetbeat. https://github.com/elastic/packetbeat/tree/mongodb

The tests work again, but I didn't look much beyond that.

**Comment by tsg** [comment]
I added a couple more commits to that branch. In short:

* Added a few more unit tests that work at module level. Not that the coverage was bad before, but I like to add tests while reviewing :-).

* Made the `send_request` and `send_response` option work like for the other protocols. This meant renaming the `mongodb.documents` in `response` (which is an analyzed field in ES, and all protocols have it) and disabling it by default (to keep the docs small and avoid indexing sensitive data).

Things that I'd still like to have a look at:

* [ ] Also hide the documents from inserts/updates by default
* [ ] Add an option to limit how many documents are indexed (often having a preview of the first 10 docs is enough)
* [ ] See if we can make the case of large responses accept some packet loss (i.e. implement the GapInStream method)

If fine with you @albanm, I'll go over this things early next week and then merge it.

File: protos/mongodb/mongodb_structs.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	"distinct",
+	"group",
+	"mapReduce",
+	"geoNear",
+	"geoSearch",
+	"geoWalk",
+	"insert",
+	"update",
+	"delete",
+	"findAndModify",
```
Comment (tsg) [inline protos/mongodb/mongodb_structs.go]: It seems that these commands show up all lower case in the protocol, so I think it's better to store them like that in this array. e.g. `findandmodify` not `findAndModify`

File: protos/mongodb/mongodb_parser.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+	query, err := d.readDocument()
+	if d.i < len(d.in) {
+		m.event["returnFieldsSelector"], err = d.readDocumentStr()
+	}
+
+	// Actual method is either a 'find' or a command passing through a query
+	if strings.HasSuffix(fullCollectionName, ".$cmd") {
+		m.method = "otherCommand"
+		for _, command := range UserCommands {
+			if _, present := query[command]; present {
```
Comment (tsg) [inline protos/mongodb/mongodb_parser.go]: Here we need to also check that the the value is a string, for example: `update: restaurants` should match, but `update: {}` should not. This is needed for the `updateandmodify` command.

**Comment by tsg** [comment]
The above commit addresses the issues raised above, I only mentioned them on the diff to make it easier to understand the changes.

**Comment by tsg** [comment]
This was merged in master already last week. Thanks again @albanm for the great contribution!

**Comment by albanm** [comment]
Thanks! I'm glad the functionality made it to master. It seems that you had quite a bit of work left to finish the work properly.

Any idea, even vague, as to when this will make it into a release ?

**Comment by tsg** [comment]
We're planning to cut the release early next week.

**Pull request closed without merging**